test/test.cpp

```diff
@@ -0,0 +1,39 @@
+#include <test.hpp>
+
+bool glob_match(const std::string& input, const std::string& pattern)
+{
+    return test::glob_match(input.begin(), input.end(), pattern.begin(), pattern.end());
+}
+
+TEST_CASE(globbing)
+{
+    EXPECT(not glob_match("ab", "a"));
+    EXPECT(not glob_match("ba", "a"));
+    EXPECT(glob_match("ab", "ab"));
+
+    // Star loop
+    EXPECT(glob_match("/foo/bar/baz/blig/fig/blig", "/foo/*/blig"));
+    EXPECT(glob_match("/foo/bar/baz/xlig/fig/blig", "/foo/*/blig"));
+    EXPECT(glob_match("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab", "a*a*a*a*a*a*a*a*b"));
+    EXPECT(glob_match("aabaabaab", "a*"));
+    EXPECT(glob_match("aabaabaab", "a*b*ab"));
+    EXPECT(glob_match("aabaabaab", "a*baab"));
+    EXPECT(glob_match("aabaabaab", "aa*"));
+    EXPECT(glob_match("aabaabaab", "aaba*"));
+    EXPECT(glob_match("aabaabqqbaab", "a*baab"));
+    EXPECT(glob_match("aabaabqqbaab", "a*baab"));
+    EXPECT(glob_match("abcdd", "*d"));
+    EXPECT(glob_match("abcdd", "*d*"));
+    EXPECT(glob_match("daaadabadmanda", "da*da*da*"));
+    EXPECT(glob_match("mississippi", "m*issip*"));
+
+    // Repeated star
+    EXPECT(glob_match("aabaabqqbaab", "a****baab"));
+    EXPECT(glob_match("abcdd", "***d"));
+    EXPECT(glob_match("abcdd", "***d****"));
+
+    // Single wildcard
+    EXPECT(glob_match("abc", "a?c"));
+}
+
+int main(int argc, const char* argv[]) { test::run(argc, argv); }
```

Comment:
```
#include <iostream>
#include <string>
#include <regex>

std::string do_regexp(const std::string & str, const std::string & rstr) {
  std::string rv;
  std::regex r(rstr);
  std::smatch m;
  std::cout << str << "  [" << rstr << "] ---> " << std::flush;
  auto match = std::regex_search(str, m, r);
  if(match) 
    std::cout << m.str();	// the matched part
  std::cout << '\n';

  return m.str();
}

int main()
{
  do_regexp("/foo/bar/baz/blig/fig/blig", "/foo/.*/blig");
  do_regexp("/foo/bar/baz/xlig/fig/blig", "/foo/.*/blig");
  do_regexp("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab", "a*a*a*a*a*a*a*a*b");
  do_regexp("aabaabaab", "a.*");
  do_regexp("aabaabaab", "a.*b.*ab");
  do_regexp("aabaabaab", "a.*baab");
  do_regexp("aabaabaab", "aa.*");
  do_regexp("aabaabaab", "aaba*");
  do_regexp("aabaabqqbaab", "a*baab");
  do_regexp("aabaabqqbaab", "a*baab");
  do_regexp("abcdd", ".*d");
  do_regexp("abcdd", ".*d*");
  do_regexp("daaadabadmanda", "da*da*da*");
  do_regexp("mississippi", "m*issip*");
  do_regexp("a big cat", "zz");
  do_regexp("a big cat", "cat*");     
  return 0;
}
```

Reply:
Please note, the `*` has to be written as `.*`, because `*` operates on something.. Thanks.

Reply:
Yea there are problem with using regex directly like this:

1) Running `./bin/my_test test_foo` will run both `test_foo` and `test_foo_with_something`, when it should only run `test_foo`.
2) The test name can't always be copied directly to the command-line, since many characters need to be escaped in the regex.

This is why I am using globbing instead of a regex as it is mostly backwards compatible with what we are doing currently.

Reply:
> Yea there are problem with using regex directly like this:
> 
> 1. Running `./bin/my_test test_foo` will run both `test_foo` and `test_foo_with_something`, when it should only run `test_foo`.

Not really. You have many ways to solve it. One is to compare the original string with the matched results. Other is to search for the correct regexp:   `do_regexp("test_foo_with_something", "test_foo$")`

> 2. The test name can't always be copied directly to the command-line, since many characters need to be escaped in the regex.

The test cases that you listed are so easy to _escape_. I of course can't comment on sequences not provided in the example. Thanks.
> 
> This is why I am using globbing instead of a regex as it is mostly backwards compatible with what we are doing currently.

Reply:
> One is to compare the original string with the matched results.

That can become ambiguous as the meaning changes if there a test with the exact name or not.

> Other is to search for the correct regexp: do_regexp("test_foo_with_something", "test_foo$")

Thats not backwards compatible. We should be able to just paste the name of the test to run it.

> The test cases that you listed are so easy to escape. I of course can't comment on sequences not provided in the example. 

Those test cases weren't added since there isn't any special characters needed to escape, but I can add some test cases like `glob_match("test.foo[gpu]", "test.foo[gpu])`, which the regex would fail. 

From an interface perspective the globbing is much simpler than a regex and it can be a drop in replacement. This is why globbing(and not a regex) is used for referencing files in the shell as well as for other things like checks for [clang tidy](https://clang.llvm.org/extra/clang-tidy/) or test names for [Catch2](https://github.com/catchorg/Catch2).

Reply:
Please add all the test cases that you think you need to address in that test file. Else this would be a discussion of hypotheticals. Thanks.

Reply:
Sure tests are added. 

Also I checked the performance and using `std::regex_search` is 100x slower than `glob_match` for the same test cases, so its also a very significant slowdown:

```
[   RUN    ] globbing
[ COMPLETE ] globbing (1.31127ms)
[   RUN    ] regex
[ COMPLETE ] regex (137.184ms)
[==========] 2 tests ran
```

And, `glob_match` is written for simplicity and not performance, it could be optimized more.

Reply:
I have provided the sample equivalent code (_with more powerful rules than that are implemented in this PR_), and the execution took just 4 milli seconds. Thus performance is not an issue. Please use the built-in super powerful standard regexp library.

Reply:
Perf was just one issue. the more important issue is usability. Sure a regex is much more powerful but introduces a lot more complexity in the interface. I dont want to have insert `^` or `$` or add in escape characters, instead of passing in `test.foo[gpu]` I need pass in `"^test\\.foo\\[gpu\\]\$"`.

This also changes the behavior of the test drivers which takes a test case name as an input. The globbing is much simpler and maintains backwards compatibility. Its also how its done in other test frameworks such as [Catch2](https://catch2-temp.readthedocs.io/en/latest/command-line.html#specifying-which-tests-to-run) and [googletest](https://google.github.io/googletest/advanced.html#running-a-subset-of-the-tests). I dont think the extra capability of regexes in this case is useful(otherwise popular test frameworks like Catch2 or googletest would use it) or worth the extra complexity(such as escaping or using `^` or `$`).